allowing dotplot to use two variables in groupby as x and y axis

zhangguy opened the pull request:
Hi,

This is for #1876 
I added a new variable `groupby_expand` to dotplot and baseplot to allow the function using two variables from groupby as x and y axis. 
Thanks.

Commit:
allowing dotplot to use two variables in groupby as x and y axis

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
- added doc and example for dotplot groupby_expand = True

- made groupby_expand and categories_order compatib

ivirshup commented:
Thanks for the PR!

While I like the idea of being able to be more flexible around how the dot plots are made, I'm not sure I like the idea of adding more argument specific behavior. That is, I don't like that passing a boolean flag changes the meaning of the values for the `var_names` and `groupby` arguments. E.g. the `group_by` is the selection of `obs` corresponding to the plots rows, while `var_names` was the variable for the columns.

What could maybe be done instead is to have an argument for column grouping. Instead of

```python
sc.pl.dotplot(adata, var_names='C1QA', groupby=['louvain', 'sampleid'], groupby_expand=True)
```

It could be more like

```python
sc.pl.dotplot(adata, var_names='C1QA', group_by='louvain', group_cols='sampleid')
```

I would ideally like a solution here to work for all the other "grouped plots" as well.

I'll write a bit more about this in the parent issue.

Commit:
- changed variable 'groupby_expand' to 'col_groups' in _dotplot and _baseplot.
- added 'col_groups' to _matrixplot.
- If given col_groups, dotplot and matrixplot will use this variables or combo of variables as the x axis.

Commit:
fix pre-commit

Commit:
Merge branch 'master' into dotplot2d
merge with master
